File: src/algorithms/xex_fde.jl
module FDE

include("sea128.jl")
using .Sea128: encrypt_sea, decrypt_sea
using Base64


function mul_alpha!(tweak::Vector{UInt8})
    carry = (tweak[16] & 0x80) != 0
    for i in 16:-1:2
        tweak[i] = ((tweak[i] << 1) & 0xFF) | (tweak[i-1] >> 7)
    end
    tweak[1] = ((tweak[1] << 1) & 0xFF)
    tweak[1] = tweak[1] ⊻ (carry ? 0x87 : 0x00)
end


function crypt_fde(key::Array{UInt8}, tweak::Array{UInt8}, input::Array{UInt8}, mode::String)
    k1, k2 = key[1:16], key[17:32]

    tweak = encrypt_sea(k2, tweak)
    text = Array{UInt8}(undef, 0)
    crypt_function = mode == "encrypt" ? encrypt_sea : decrypt_sea

    for i in 1:16:length(input)
        block = input[i:i+15] .⊻ tweak
        append!(text, (crypt_function(k1, block) .⊻ tweak)) 
        mul_alpha!(tweak)
    end
    return text
end

function encrypt_fde(key::Array{UInt8}, tweak::Array{UInt8}, input::Array{UInt8})
    return crypt_fde(key, tweak, input, "encrypt")
end

function decrypt_fde(key::Array{UInt8}, tweak::Array{UInt8}, input::Array{UInt8})
    return crypt_fde(key, tweak, input, "decrypt")
end

end

File: src/algorithms/xex_fde.cpp
// src/algorithms/xex_fde.cpp
#include "xex_fde.h"
#include "sea128.h"
#include <vector>
#include <string>
#include <stdexcept>

// Helper function to multiply tweak by alpha
void mul_alpha(std::vector<uint8_t>& tweak){
    bool carry = (tweak[15] & 0x80) !=0;
    for(int i=15; i>0; --i){
        tweak[i] = (tweak[i] <<1) | (tweak[i-1] >>7);
    }
    tweak[0] = (tweak[0] <<1);
    if(carry){
        tweak[0] ^= 0x87;
    }
}

std::vector<uint8_t> crypt_fde(const std::vector<uint8_t>& key, std::vector<uint8_t> tweak, const std::vector<uint8_t>& input, const std::string& mode){
    if(key.size() !=32){
        throw std::invalid_argument("Key must be 32 bytes");
    }
    std::vector<uint8_t> k1(key.begin(), key.begin()+16);
    std::vector<uint8_t> k2(key.begin()+16, key.end());

    tweak = encrypt_sea(k2, tweak);

    std::vector<uint8_t> output;
    std::string crypt_mode = mode == "encrypt" ? "encrypt_sea" : "decrypt_sea";

    for(size_t i=0; i < input.size(); i +=16){
        std::vector<uint8_t> block(input.begin()+i, input.begin()+std::min(i+16, input.size()));
        // XOR with tweak
        for(size_t j=0; j<block.size(); ++j){
            block[j] ^= tweak[j];
        }
        // Encrypt/decrypt
        std::vector<uint8_t> crypt_block;
        if(mode == "encrypt"){
            crypt_block = encrypt_sea(k1, block);
        }
        else{
            crypt_block = decrypt_sea(k1, block);
        }
        // XOR with tweak
        for(size_t j=0; j<crypt_block.size(); ++j){
            crypt_block[j] ^= tweak[j];
        }
        // Append to output
        output.insert(output.end(), crypt_block.begin(), crypt_block.end());
        // Multiply tweak by alpha
        mul_alpha(tweak);
    }
    return output;
}

std::vector<uint8_t> encrypt_fde(const std::vector<uint8_t>& key, const std::vector<uint8_t>& tweak, const std::vector<uint8_t>& input){
    std::vector<uint8_t> tweak_copy = tweak;
    return crypt_fde(key, tweak_copy, input, "encrypt");
}

std::vector<uint8_t> decrypt_fde(const std::vector<uint8_t>& key, const std::vector<uint8_t>& tweak, const std::vector<uint8_t>& input){
    std::vector<uint8_t> tweak_copy = tweak;
    return crypt_fde(key, tweak_copy, input, "decrypt");
}
